### ISSUE TITLE
Allow update also of (status) values

### ISSUE THREAD

**Issue by robertfeldt** [timholy/ProgressMeter.jl#40]
Thanks for ProgressMeter; very useful. For long-running computations (such as for example in optimization) it would be nice to allow a dict (or some other way) to specify current values which can then be reported on the lines below the progress bar? For example, in the optimiation, when want to know the overall progress in terms of time (ETA) but also would like to know the objective value of the current best solution found. If these are just printed to STDOUT it would mess up the reporting from ProgressMeter so it seems there needs to be some support for this. Has this been considered as a future extension? If not, if this is something you could accept as a PR I might take a stab at it.

**Comment by timholy** [comment]
Did you see the `ProgressThresh` type?

**Comment by robertfeldt** [comment]
I did, yes, but my impression is that it is for cases when I want to affect the progress with said value. In my case I'm trying to find a way to present additional information about the progress but which should not determine the actual progress. Does it make sense? ;)

**Comment by robertfeldt** [comment]
Example: At one point in an optimization run I may have done 50 out of 100 iterations and I want to present the current minima found say:

progress  50% ... ETA: 00:14:12
  current minima: 0.3452

It would be quite nice if it can also model the EVA (Estimated Value at Arrival) using information about the recent progress on the value or something. But most important is to be able to show additional information without messing up the progress meter output.

**Comment by timholy** [comment]
It's intended for cases where convergence is determined by value rather than number of iterations---for example, an optimization in which convergence occurs when the function value changes by less than `tol`. It might take 3 or 3000 iterations to reach `tol`, and you can't predict this in advance. (Please do feel free to improve the documentation, given that it was not sufficiently clear.)

It looks like you have both a deterministic number of iterations and a value you want to report back. That's basically a hybrid of the two, so I'd guess that you'd indeed want a new type. I'd certainly consider a PR implementing this.

**Comment by robertfeldt** [comment]
Ok, I have a busy week but will take a stab at this after that. Basic idea is for an optional argument to `next!` and `update!` like so:

```julia
p = ProgressMeter(100, 1, "progress  ", 70)
...
next!(p; minima = current_best.fitness, meanfitness = mean(map(i -> i.fitness, individuals))) # for some values of interest, here lets say minima is 0.25 and meanfitness is 0.567
````
which will print after 50 iterations:
````
progress  50% ||||||||||...bar here...|||||                     ETA: 00:07:13
  minima: 0.25
  meanfitness: 0.567
````
Possibly with some better formatting/colouring to indicate if its increasing or decreasing. The way to implement it would be to just save the number of extra lines that was printed below the progress bar and then print out that same number of `"\033[F"` before printing the progress bar line.

Comments?

**Comment by timholy** [comment]
That could work. But, what if the variable you want to print happens to have the same name as one of the existing keywords? Seems like it might be better to use something like `next!(p; showvalues = Dict(:minima=>current_best.fitness, :meanfitness=>...))`.

Seems like it would also be important to test what happens when the user supplies `showvalues` on some calls but not on others within the same loop.

**Comment by robertfeldt** [comment]
Yes collecting them in one arg is better. We can also allow an array of tuples if one wants to control the order explicitly. By looping over the arg with `for (name, value) in showvalues` we should be able to cover both. Ok, I'll explore it unless you have more concerns.

**Comment by robertfeldt** [comment]
Ok I did a PR, please see: https://github.com/timholy/ProgressMeter.jl/pull/41

**Comment by ahwillia** [comment]
Should this be closed as done/implemented? Or is there more work to be done?

Maybe a minimal example should be added to the README before closing?

**Comment by robertfeldt** [comment]
From my perspective it is done and I think Tim integrated it. Maybe lift one of the tests as an example into the README?

**Comment by timholy** [comment]
Yes, thanks for noticing. Agreed that a README example would be nice.

**Comment by timholy** [comment]
Closed by #41 and #55
